templates/SynapseExternalBucket.yaml

```diff
@@ -123,6 +136,111 @@ Resources:
+    Type: "AWS::Lambda::Function"
+    Condition: CreateIPAddressRestrictionLambda
+    Properties:
+      Description: Dynamically updates the external bucket policy to restrict downloading only IP addresses listed by AWS.
+      Runtime: python3.6
+      Handler: "index.handler"
+      Role: !GetAtt ExternalBucketGroupPolicyUpdateRole.Arn
+      Timeout: 10
+      Code:
+        ZipFile: !Sub
+        - |
+          import json
+          import boto3
+          from botocore.vendored import requests
+
+          bucket_name = '${BucketName}'
+          region = '${Region}'
+          policy_statement_id="DenyGetObjectForNonMatchingIp"
+
+          def handler(event, context):
+            s3_client = boto3.client('s3')
+
+            #generate new policy statement based on data from AWS
+            ip_ranges = requests.get('https://ip-ranges.amazonaws.com/ip-ranges.json').json()['prefixes']
+            region_ip_addresses = [item['ip_prefix'] for item in ip_ranges if (item["service"] == "AMAZON" and item["region"] == region)]
+            
+            new_ip_policy_statement = {'Sid': policy_statement_id,
+                                       'Effect': 'Deny',
+                                       'Principal': '*',
+                                       'Action': 's3:GetObject',
+                                       'Resource': 'arn:aws:s3:::'+bucket_name+'/*',
+                                       'Condition': {'NotIpAddress': {'aws:SourceIp': region_ip_addresses}}}
+
+
+            #get current bucket_policy from the s3 bucket
+            bucket_policy = json.loads(s3_client.get_bucket_policy(Bucket=bucket_name)['Policy'])
+
+            #remove the old IP address policy statement and add new
+            bucket_policy['Statement'] = [statement for statement in bucket_policy['Statement'] if (policy_statement_id != statement.get("Sid"))]
+            bucket_policy['Statement'].append(new_ip_policy_statement)
+
+
+            s3_client.put_bucket_policy(Bucket=bucket_name, Policy=json.dumps(bucket_policy))
+
+        - { BucketName: !If [EnableEncryption, !Ref SynapseEncryptedExternalBucket, !Ref SynapseExternalBucket], Region: !Ref "AWS::Region" }
+  ExternalBucketGroupPolicyUpdateSNSSubscription:
+    Condition: CreateIPAddressRestrictionLambda
+    Type: AWS::SNS::Subscription
+    Properties:
+      TopicArn: 'arn:aws:sns:us-east-1:806199016981:AmazonIpSpaceChanged'
+      Protocol: lambda
+      Endpoint: !GetAtt ExternalBucketGroupPolicyUpdateLambda.Arn
+  ExternalBucketGroupPolicyUpdateLambdaSNSInvokePermission: # for sns to trigger group policy update lambda
+    Condition: CreateIPAddressRestrictionLambda
+    Type: AWS::Lambda::Permission
+    Properties:
+      FunctionName: !GetAtt ExternalBucketGroupPolicyUpdateLambda.Arn
+      Action: 'lambda:InvokeFunction'
+      SourceArn: 'arn:aws:sns:us-east-1:806199016981:AmazonIpSpaceChanged'
+      Principal: sns.amazonaws.com
+  ExternalBucketGroupPolicyUpdateTimer:
```

Comment:
why do you need 2 separate triggers for updating?

Reply:
Yeah the SNS trigger should be enough for updates.
I added a new CloudFormation Custom Resource to make the lambda apply the group policy when during the execution of the template.